### ISSUE TITLE
Add capability to set project root with environment variable

### ISSUE THREAD

**Issue by benjumanji** [ul/kak-lsp#182]
Some build systems use recursively placed build files. In large scala projects it isn't weird to see `build.sbt` in multiple places, but the scala language server really needs to be started at the parent build.sbt. I don't think it's reasonable to expect kak-lsp to be able to support this, so allowing people to just _tell_ you where the root is would be a helpful feature. 

Other weirdo build systems that might benefit: bazel / tup / pants / waf

I'd also be happy to work on this.

**Comment by ul** [comment]
Yes, PR is welcome for this feature!

**Comment by benjumanji** [comment]
Project Root selection
==================

## Problem
Current solution allows the specification of a set of markers per filetype for
the project root. On encountering a new buffer we walk from the buffer towards
the filesystem root checking at each directory if we have a match for a marker.
We stop at the first marker. This works well build systems with a single root
file but for systems that may scatter their build files all over the project
directory (in some extreme cases in _every_ folder c.f.
[tup](http://gittup.org).

## Solution
Provide a hook to allow projects roots to be explicitly designated on a
per-filetype basis. This patch uses environment variables. For maximum
flexibility the following scheme is used: on new buffer detect filetype, and
look for all environment variables prefixed with `KAK_LSP_%FILETYPE%*` and add
the value of each to a set of roots and proceed as before, first match wins.
This will allow for both regular repos and monorepos browsed from the root to
specify subprojects accurately. Consider the following tree:
```
~/src
├── bar
│   ├── build.sbt
│   └── ext
│       ├── sub1
│       │   ├── build.sbt
│       │   └── src
│       │       └── main
│       │           └── scala
│       └── sub2
│           ├── build.sbt
│           └── src
│               └── main
│                   └── scala
└── foo
    ├── build.sbt
    └── ext
        ├── sub1
        │   ├── build.sbt
        │   └── src
        │       └── main
        │           └── scala
        └── sub2
            ├── build.sbt
            └── src
                └── main
                    └── scala
```
Then an enviroment with `KAK_LSP_SCALA_FOO="$HOME/src/foo"` and
`KAK_LSP_SCALA_BAR="$HOME/src/bar"` would allow both sets of files to
successfully be associated with the correct build. Of course users that only
browse from single sources will be catered too as well and need not suffix at
all. Note that the `FOO` and `BAR` suffixes are _arbitrary_ and have no meaning
when it comes to matching.

**Comment by benjumanji** [comment]
see above for a design proposal. I wrote it out as an md and was going to PR it but I wasn't sure if you want a folder of rfcs in your source repo or not.

**Comment by ul** [comment]
I never thought about having RFCs directory in the repo as kak-lsp didn't look large enough to have such a disciplined process on top of free-form discussion in issues though keeping RFCs could be a good investment into the documenting design choices.

Regarding the proposal itself, it looks good to me. The only thing which caught my eye is that `KAK_LSP_%FILETYPE%*` naming scheme seems to be too generic — what if in the future we will decide to make another option to be controlled by env var based on filetype? Something like `KAK_LSP_PROJECT_ROOT_%FILETYPE%*` looks a bit better to me despite being wordy (if it feels too wordy `PROJECT` could be dropped I guess.)

**Comment by benjumanji** [comment]
Ah yes. My proposal does steal too much namespace. I think stuffing project root in there is fine. It's not like anyone will be typing these out over and over. Expect a PR shortly.

**Comment by ul** [comment]
I created a wiki entry with a slight modification of your RFC as documentation for the feature: https://github.com/ul/kak-lsp/wiki/Project-root-selection

**Comment by benjumanji** [comment]
nice!
